src/main/java/org/magicdgs/readtools/utils/read/stats/engine/ProperStatWindowEngine.java

```diff
@@ -0,0 +1,275 @@
+            // TODO: if using IntervalsSkipList it might be easier by querying needed ones
+            w.addRead(read);
+        }
+    }
+
+    /**
+     * Flushes the queue of windows up to the provided read.
+     *
+     * @param read next read to be added.
+     * @return windows overlapping with the read.
+     */
+    private List<ProperStatWindowCalculator> flushUpToRead(final GATKRead read) {
+        // edge case where the windows are already finalized
+        if (windowsPerContig.isEmpty()) {
+            return Collections.emptyList();
+        }
+
+        // store the first key
+        String firstKey = windowsPerContig.firstKey();
+        // only if the read is mapped, otherwise does not have a contig
+        if (!read.isUnmapped()) {
+            logger.debug("Flush up to {} ({} contigs)", read::toString, windowsPerContig::size);
+            final String readContig = read.getContig();
+            // remove the contigs that are before the current read
+            while (!windowsPerContig.isEmpty() && windowsPerContig.comparator().compare(firstKey, readContig) < 0) {
+                logger.debug("Output contig {}", firstKey);
+                windowsPerContig.remove(firstKey).forEach(this::maybeWriteWindow);
+                firstKey = (windowsPerContig.isEmpty()) ? null : windowsPerContig.firstKey();
+            }
+            logger.debug("After flush {} contigs remain", windowsPerContig::size);
+        }
+        // TODO: performance improvement - here we can also flush the List<ProperStatWindowCalculator>
+        // TODO: if the list is sorted, we should check for overlapping windows with the left-most coordinate (read or mate)
+
+        return (firstKey == null) ? Collections.emptyList() : windowsPerContig.get(firstKey);
+    }
+
+    /**
+     * Writes the header for the output file.
+     */
+    private void writeHeader() {
+        // the column names from the first window is the one to take into account
+        writer.println("window\t" + Utils.join("\t", columnNames));
+    }
+
+    /**
+     * Writes the window to disk if it is requested (e.g., more than 0 or print-all is included)
+     *
+     * @param win the window to be printed.
+     */
+    private void maybeWriteWindow(final ProperStatWindowCalculator win) {
+        final TableFeature feature = win.toTableFeature();
+        if (printAll || Integer.valueOf(feature.get("total")) != 0) {
+            writeWindow(feature);
+        }
+    }
+
+    /**
+     * Writes the feature record.
+     *
+     * @param feature the feature to be written.
+     */
+    private void writeWindow(final TableFeature feature) {
+        if (logger.isWarnEnabled() ) {
+            final int missing = Integer.valueOf(feature.get("missing"));
+            if (missing != 0) {
+                logger.warn("{} missing pairs for {}", missing, IntervalUtils.locatableToString(feature));
+            }
+        }
+        writer.println(feature.toString());
```

Comment:
Add TODO for a different formatting: I don't think that it is good to rely on the `toString()` method

Reply:
It is ok as the contract of `TableFeature` says that it is the tab-delimited format. But the format might change anyway due to https://github.com/magicDGS/ReadTools/issues/459